File: trip-kit-booking-ui/src/main/java/com/skedgo/android/tripkit/booking/ui/viewmodel/DisposableViewModel.java
package com.skedgo.android.tripkit.booking.ui.viewmodel;

import android.support.annotation.CallSuper;
import android.support.annotation.NonNull;

import rx.Observable;
import rx.subjects.PublishSubject;

public abstract class DisposableViewModel {
  private final PublishSubject<Void> _onDispose = PublishSubject.create();

  public final @NonNull Observable<Void> onDispose() {
    return _onDispose.asObservable();
  }

  @CallSuper public void dispose() {
    _onDispose.onNext(null);
  }
}

File: trip-kit-booking-ui/src/main/java/com/skedgo/android/tripkit/booking/ui/viewmodel/DisposableViewModel.kt
package com.skedgo.android.tripkit.booking.ui.viewmodel

import androidx.annotation.CallSuper
import io.reactivex.Observable
import io.reactivex.subjects.PublishSubject

enum class Nothing {
  INSTANCE
}

abstract class DisposableViewModel {
  private val _onDispose = PublishSubject.create<Nothing>()

  fun onDispose(): Observable<Nothing> {
    return _onDispose.hide()
  }

  @CallSuper fun dispose() {
    _onDispose.onNext(Nothing.INSTANCE)
  }
}
